### DISCUSSION TITLE
Starting a testcontainer with pre-existing data dir/volume

### DISCUSSION THREAD

**Discussion by jubui** [testcontainers/testcontainers-java#8734] [Q&A]
I am trying to simulate restart behavior (there is no 'restart' testcontainer API provided, right?) and to do so, I've been trying to start up a container, stop it, then create a new container that mounts the data dir of the first container. I've tried doing so with the following approaches:

- withFileSystemBind, but encountered root owner issues. My host dir would have a non root UID and then would bind so that the user in the container becomes root
- withCopyFileToContainer: Stopping the container creates an archive/tar of my data directory, then put that archive on the host dir, then extract it on the host dir, and then `withCopyFileToContainer(MountableFile.forHostPath(<dir>)...` but that doesn't work well because on macs you can't preserve the UID if you don't have a user of that UID and  unless you `sudo` a bunch of commands
- Startables: was hoping to use a Startable to put the archive on the container and then extract it BEFORE the container start() method is called. Issue here seemed to be that the startable code doesn't get run on the container you want to start with
- An above option and additionally executing a script in the container that `chown`s the entire volume to the right user? it doesn't seem I can do this with docker's entrypoint + command

None of these options really worked and mainly, I think, because my host machine doesn't have the UID of the user to whom the data directory belongs in my system.

Questions:
1) What is the best way to supply a data directory to my container BEFORE it starts? and avoid the UID issue
2) Is there a recommended way to simulate restarts with testconatiners?

**Comment by eddumelendez** [comment]
Hi,

> What is the best way to supply a data directory to my container BEFORE it starts? and avoid the UID issue

I'd say is using `liquibase` or `flyway`. Testcontainers Database implementation also supports `withInitScript`. Also, many images supports a initialization directory, where all scripts can be copied and then the container will be in charge of the execution.

Another approach is running an empty database and then populate it with one of the approaches mentioned above and then commit the image. So, next time you can use the new image instead. See [this code snippet](https://github.com/testcontainers/testcontainers-java/blob/main/modules/ollama/src/main/java/org/testcontainers/ollama/OllamaContainer.java#L60-L75)

> Is there a recommended way to simulate restarts with testcontainers?

Do you really want to test restarts or want to test network failures? If so, then you can use [Toxiproxy](https://testcontainers.com/modules/toxiproxy/)
